src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -341,5 +365,110 @@ std::string GcsResourceManager::ToString() const {
+            cluster_resource_state_version_);
+
+  reply->set_last_seen_autoscaler_state_version(last_seen_autoscaler_state_version_);
+  reply->set_cluster_resource_state_version(cluster_resource_state_version_);
+
+  GetNodeStates(reply);
+  GetPendingResourceRequests(reply);
+  GetPendingGangResourceRequests(reply);
+  GetClusterResourceConstraints(reply);
+
+  // QQ: should we use GCS_RPC_SEND_REPLY
+  // That would add another field in the reply, and introduce dep of `GcsStatus`.
+  send_reply_callback(ray::Status::OK(), nullptr, nullptr);
+}
+
+void GcsResourceManager::GetPendingGangResourceRequests(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  // TODO
+  return;
+}
+
+void GcsResourceManager::GetClusterResourceConstraints(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  // TODO
+  return;
+}
+
+void GcsResourceManager::GetPendingResourceRequests(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  std::unordered_map<google::protobuf::Map<std::string, double>, rpc::ResourceDemand>
+      aggregate_load;
+
+  for (const auto &usage : node_resource_usages_) {
+    // Aggregate the load reported by each raylet.
+    FillAggregateLoad(usage.second, &aggregate_load);
+  }
+
+  for (const auto &[shape, demand] : aggregate_load) {
+    // QQ: should backlog size to be included here?
+    auto num_pending = demand.num_infeasible_requests_queued();
+    if (num_pending > 0) {
+      auto pending_req = reply->add_pending_resource_requests();
+      pending_req->set_count(num_pending);
+      auto req = pending_req->mutable_request();
+      req->mutable_resources_bundle()->insert(shape.begin(), shape.end());
+    }
+  }
+}
+
+void GcsResourceManager::GetNodeStates(
+    rpc::autoscaler::GetClusterResourceStateReply *reply) {
+  for (auto &[_node_id, node_state] : node_states_) {
+    auto node_state_proto = reply->add_node_states();
+    node_state_proto->CopyFrom(node_state);
+
+    // QQ: There's node_resource_usages_ and cluster resource manager for the node
+    // resources? Why's seemingly 2 sources of truth?
+    //
+    // const auto &node_resource_data =
+    //     node_resource_usages_.at(NodeID::FromBinary(node_state.node_id()));
+
+    auto const &node_resource_data = cluster_resource_manager_.GetNodeResources(
+        scheduling::NodeID(node_state.node_id()));
+
+    // Copy resource available
+    const auto &available = node_resource_data.available.ToResourceMap();
+    node_state_proto->mutable_available_resources()->insert(available.begin(),
+                                                            available.end());
+
+    // Copy total resources
+    const auto &total = node_resource_data.total.ToResourceMap();
+    node_state_proto->mutable_total_resources()->insert(total.begin(), total.end());
+
+    // TODO: support dynamic labels.
+  }
+}
+
+void GcsResourceManager::RemoveNodeState(const NodeID &node_id) {
+  node_states_.erase(node_id);
```

Comment:
do we need to report DEAD node information? If we remove it here, how do we report them
? I found DEAD status from the NodeStatus

Reply:
Oh yeah, that's another question I had in mind. 

I think we will not report DEAD nodes actually:
- From Ray's perspetice, a node is a DEAD when it unregistered with GCS (OnNodeDead())
- From autoscaler's perspective, a node is DEAD if it's not included in the `node_states` 

But I think the alternative of keeping DEAD nodes info could also work. 

@scv119

Reply:
it'd be nice to report the dead status to autoscaler for health checking purpose. I believe it will make reconciliation logic simpler.

concretely the case i'm thinking is:

> a node doesn't show up in gcs, but it shows up as ALIVE in node provider. 
in this case, we can't tell if the node is DEAD, or the node is actually pending start up from GCS, due to consistency issue.

however, with the extra dead info it will be easy to tell: we only decide the node is DEAD if gcs reports the node is DEAD.

agreed it's possible to remove this state and we can still induct the state of node, but probably a bit more complex.